testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,538 @@
+  }
+
+  private boolean isStargateEnvInfo(ParameterContext parameterContext) {
+    return parameterContext
+        .getParameter()
+        .getType()
+        .isAssignableFrom(StargateEnvironmentInfo.class);
+  }
+
+  @Override
+  public boolean supportsParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    return isStargateConnectionInfo(parameterContext) || isStargateEnvInfo(parameterContext);
+  }
+
+  @Override
+  public Object resolveParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    if (isStargateEnvInfo(parameterContext)) {
+      return container();
+    } else if (isStargateConnectionInfo(parameterContext)) {
+      return container().nodes.get(0);
+    }
+
+    throw new IllegalStateException("Unknown parameter: " + parameterContext);
+  }
+
+  private static boolean isDebug() {
+    return getRuntimeMXBean().getInputArguments().toString().contains("-agentlib:jdwp");
+  }
+
+  @StargateSpec
+  private static final class DefaultSpecHolder {
+
+    private static StargateSpec get() {
+      return DefaultSpecHolder.class.getAnnotation(StargateSpec.class);
+    }
+  }
+
+  private static class Container implements StargateEnvironmentInfo {
+
+    private final UUID id = UUID.randomUUID();
+    private final int instanceNum = stargateInstanceSeq.getAndIncrement();
+    private final ClusterConnectionInfo backend;
+    private final StargateSpec spec;
+    private final StargateParameters parameters;
+    private final List<Node> nodes = new ArrayList<>();
+
+    private Container(
+        ClusterConnectionInfo backend, StargateSpec spec, StargateParameters parameters)
+        throws Exception {
+      this.backend = backend;
+      this.spec = spec;
+      this.parameters = parameters;
+
+      Env env = new Env(spec.nodes());
+      for (int i = 0; i < spec.nodes(); i++) {
+        nodes.add(new Node(i, instanceNum, backend, env, parameters));
+      }
+    }
+
+    private void start() {
+      for (Node node : nodes) {
+        node.start();
+      }
+
+      for (Node node : nodes) {
+        node.awaitReady();
+      }
+    }
+
+    private void stop() {
+      for (Node node : nodes) {
+        node.stopNode();
+      }
+
+      for (Node node : nodes) {
+        node.awaitExit();
+      }
+    }
+
+    private boolean matches(
+        ClusterConnectionInfo backend, StargateSpec spec, StargateParameters parameters) {
+      return this.backend.id().equals(backend.id())
+          && this.spec.equals(spec)
+          && this.parameters.equals(parameters);
+    }
+
+    @Override
+    public String id() {
+      return id.toString();
+    }
+
+    @Override
+    public List<? extends StargateConnectionInfo> nodes() {
+      return nodes;
+    }
+  }
+
+  private static class Node extends Thread implements StargateConnectionInfo {
```

Comment:
It feels redundant to start our own thread, when commons-exec is already doing things asynchronously.
We could use `execute(CommandLine, ExecuteResultHandler)`, and do `exit.countDown()` in the handler.

Kind of similar to what you're already doing for `ready`.

Reply:
`DefaultExecutor` will create a new thread in that case anyway, but we won't control the thread's name :thinking: 

We can avoid creating a thread with `CommandLauncher`, but then we'd have to do a lot more work managing I/O streams... I'm not sure it's worth the trouble. We do not have that many SG nodes anyway... WDYT?

Reply:
`DefaultExecutor` already does everything asynchronously. The only reason why you're holding a thread is to wait until the command has exited, and run those 3 instructions:
```java
LOG.info("Stargate node {} existed with return code {}", nodeIndex, retValue);
FileUtils.deleteDirectory(cacheDir);
exit.countDown();
```
What I'm saying is, you can move that into an `ExecuteResultHandler` and call the async `execute()` variant. You don't need your own thread anymore, just launch the command and forget, the callback will be invoked when the command returns.

Reply:
yes, but if we call `execute(CommandLine, ExecuteResultHandler)` a thread will be created by that method anyway. So the only difference would be in who creates that thread... Did I miss something else?

Reply:
I think I found a way to set our own thread name with `execute(CommandLine, ExecuteResultHandler)`... I'll switch to that.

Reply:
> if we call `execute(CommandLine, ExecuteResultHandler)` a thread will be created by that method anyway. So the only difference would be in who creates that thread... Did I miss something else?

I hadn't looked at the commons-exec code, I thought it always started a thread and the synchronous execute was just a blocking wrapper. But it's actually the other way around, so fair enough.

So my only argument is to not reinvent the wheel if the library already does it. I don't think the thread name is hugely important, but if you've found a way to change it then all good 👍

Reply:
I changes this code to use execution listeners.